presto-main/src/main/java/com/facebook/presto/server/GracefulShutdownHandler.java

```diff
@@ -0,0 +1,129 @@
+import static java.util.concurrent.Executors.newSingleThreadScheduledExecutor;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static java.util.concurrent.TimeUnit.MINUTES;
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+public class GracefulShutdownHandler
+{
+    private static final Logger log = Logger.get(GracefulShutdownHandler.class);
+    private static final Duration GRACE_PERIOD = new Duration(2, MINUTES);
+    private static final Duration LIFECYCLE_STOP_TIMEOUT = new Duration(30, SECONDS);
+
+    private final ScheduledExecutorService shutdownHandler = newSingleThreadScheduledExecutor(threadsNamed("shutdown-handler-%s"));
+    private final ExecutorService lifeCycleStopper = newSingleThreadExecutor(threadsNamed("lifecycle-stopper-%s"));
+    private final TaskManager sqlTaskManager;
+    private final LifeCycleManager lifeCycleManager;
+    private final ShutdownFlag shutdownFlag;
+
+    @Inject
+    public GracefulShutdownHandler(TaskManager sqlTaskManager, LifeCycleManager lifeCycleManager, ShutdownFlag shutdownFlag)
+    {
+        this.sqlTaskManager = requireNonNull(sqlTaskManager, "sqlTaskManager is null");
+        this.lifeCycleManager = requireNonNull(lifeCycleManager, "lifeCycleManager is null");
+        this.shutdownFlag = requireNonNull(shutdownFlag, "shutdownFlag is null");
+    }
+
+    public void requestShutdown()
+    {
+        log.info("Shutdown requested");
+
+        shutdownFlag.set();
+
+        //wait for a grace period to start the shutdown sequence
+        shutdownHandler.schedule(() -> {
+            List<TaskInfo> allTasks = sqlTaskManager.getAllTaskInfo();
+            CountDownLatch countDownLatch = new CountDownLatch(allTasks.size());
+
+            // register state change listeners to all tasks on this worker
+            // a task may already be done when the listener is registered, but it's
+            // safe as the listener will be immediately called for a task in a terminal state
+            for (TaskInfo taskInfo : allTasks) {
+                log.info("Adding a state change listener to task %s", taskInfo);
+                sqlTaskManager.addStateChangeListener(taskInfo.getTaskId(), newState -> {
+                    if (newState.isDone()) {
+                        countDownLatch.countDown();
+                    }
+                });
+            }
+
+            log.info("Waiting for all tasks to finish");
+
+            //TODO do we want a MAX_TASK_WAIT or want to wait until all tasks are done no matter how long they take?
+            try {
+                countDownLatch.await();
+            }
+            catch (InterruptedException e) {
+                log.warn("Interrupted while waiting for all tasks to finish");
+                currentThread().interrupt();
+            }
+
+            // wait for another grace period after all tasks complete
+            //TODO I am not sure whether we need this. Can there be any in-flight messages from the
```

Comment:
I am not sure whether we need this. Can there be any in-flight messages from the coordinator to the worker after a task is done?